File: python/ohce/ui.py
from ohce import reverse


class ConsoleInteractor:
    def read_input(self):
        return input()

    def print_message(self, message):
        print(message)


class UI:
    def __init__(self):
        self.interactor = ConsoleInteractor()

    def main_loop(self):
        while True:
            input = self.interactor.read_input()
            if input == "quit":
                break
            reversed = reverse(input)
            self.interactor.print_message(reversed)
            if reversed == input:
                self.interactor.print_message("That was a palindrome!")

File: ts/ui.ts
import * as readline from 'node:readline/promises'
import { reverse } from './ohce'

class ConsoleInteractor {
  reader: readline.Interface

  constructor() {
    this.reader = readline.createInterface(
      { input: process.stdin, output: process.stdout }
    )
  }

  async readInput(): Promise<string> {
    const res = await this.reader.question("> ")
    return res
  }

  close() {
    this.reader.close()
  }

  printMessage(message: string) {
    console.log(message)
  }
}

export default class UI {
  interactor: ConsoleInteractor

  constructor() {
    this.interactor = new ConsoleInteractor()
  }

  async mainLoop() {
    while (true) {
      const input = await this.interactor.readInput()
      if (input === 'quit') {
        this.interactor.close()
        break
      }
      const reversed = reverse(input)
      this.interactor.printMessage(reversed)
      if (input === reversed) {
        this.interactor.printMessage('That was a palindrome!')
      }
    }
  }
}
